src/main/java/edu/ucla/library/prl/harvester/handlers/UpdateJobHandler.java

```diff
@@ -139,14 +142,25 @@ Future<UpdateResponse> updateSolr(final Tuple3<Job, Job, Institution> aData) {
     /**
      * @param anOldJob A Job representing an original state
      * @param aNewJob A Job representing a new, updated state
-     * @return True if new job has sets not contained in old job
+     * @return Future that resolves to true if new job has sets not contained in old job
      */
-    private boolean hasNewSets(final Job anOldJob, final Job aNewJob) {
-        return getDifference(aNewJob.getSets(), anOldJob.getSets()).isEmpty() ||
-                !anOldJob.getSets().isEmpty() && aNewJob.getSets().isEmpty() &&
-                        (getDifference(OaipmhUtils.listSets(myVertx, anOldJob.getRepositoryBaseURL(),
-                                myOaipmhClientHttpTimeout, myHarvesterUserAgent).map(OaipmhUtils::getSetSpecs).result(),
-                                anOldJob.getSets()).isEmpty());
+    //PDM doesn't know Future|Promise<Boolean> eventually becomes boolean, and boolean-style names make code clearer
+    @SuppressWarnings("PMD.LinguisticNaming")
+    private Future<Boolean> hasNewSets(final Job anOldJob, final Job aNewJob) {
+        final Promise<Boolean> hasNew = Promise.promise();
+
+        OaipmhUtils.listSets(myVertx, anOldJob.getRepositoryBaseURL(), myOaipmhClientHttpTimeout, myHarvesterUserAgent)
+                .onSuccess(sets -> {
+                    final List<String> setSpecs = OaipmhUtils.getSetSpecs(sets);
+                    final boolean simpleDiff = getDifference(aNewJob.getSets(), anOldJob.getSets()).isEmpty();
```

Comment:
If, for example, the old job specified `sets = ["set1"]` and the new job `sets = ["set1", "set2"]`: as written, `simpleDiff` evaluates to `false`, which is incorrect because `set2` was added.

I would recommend trying something similar to the four-branch conditional in `updateSolr`, conditionally calling `OaipmhUtils#listSets` and then performing a comparison between the proper lists. I think that would make it easier to clearly express the correct algorithm.

Reply:
Now I'm lost... I assume you're talking about this segment of `updateSolr`:
```
        if (!oldJobSets.isEmpty() && !newJobSets.isEmpty()) {
            // Still selective harvesting, althought the list of sets may have changed, so determine that
            getActualOldJobSets = Future.succeededFuture(oldJobSets);
        } else if (oldJobSets.isEmpty() && !newJobSets.isEmpty()) {
            // From non-selective harvesting to selective, so it's very likely that there are sets to remove
            // Must query OAI-PMH repository in order to get the sets belonging to the old job
            getActualOldJobSets = OaipmhUtils
                    .listSets(myVertx, oldJob.getRepositoryBaseURL(), myOaipmhClientHttpTimeout, myHarvesterUserAgent)
                    .map(OaipmhUtils::getSetSpecs);
            // TODO: make it impossible to change the base URL
        } else if (!oldJobSets.isEmpty() && newJobSets.isEmpty()) {
            // From selective harvesting to non-selective, so nothing to remove
            return Future.succeededFuture();
        } else {
            // Still harvesting entire repository, so nothing to remove
            return Future.succeededFuture();
        }

```Because I don't see how any of those get me to determining if there are new sets to process.

Reply:
The blocks would have to be adjusted to this particular task, for example:

```java
if (!oldJobSets.isEmpty() && !newJobSets.isEmpty()) {
    // Still selective harvesting, although the list of sets may have changed, so calculate difference between the lists on each job
    getActualNewJobSets = Future.succeededFuture(newJobSets);
} else if (oldJobSets.isEmpty() && !newJobSets.isEmpty()) {
    // From non-selective harvesting to selective (i.e., harvesting a subset of the repository's sets), so we know that none have been added
    return Future.succeededFuture();
} else if (!oldJobSets.isEmpty() && newJobSets.isEmpty()) {
    // From selective harvesting to non-selective, so query OAI-PMH repository to determine which additional sets (if any) should be harvested

    // FIXME
    getActualNewJobSets = null;
} else {
    // Still harvesting entire repository, so nothing to add
    return Future.succeededFuture();
}
```

Reply:
Are you rewriting the scope/objective of the ticket? Because none of that seems to be determining a simple boolean of whether there's new sets in the update data.

Reply:
No, this is all in scope, it's just that determining if an updated job has new sets to harvest is more difficult than simply looking at the local Job objects, since "all sets" is represented as the empty list (and thus the OAI-PMH repo needs to be queried in order to retrieve the _actual_ list to calculate the difference with). And I'm recommending the above framework for the correct algorithm for making this determination.

Reply:
I did fail to remind you of the "all sets in repo = empty list" representation from the outset, which was my bad. But hopefully I've made up for that by providing these code snippets.

Reply:
Oh, sorry -- the code snippet above should have had `return Future.succeededFuture(false);` earlier. Corrected now.

Reply:
Lemme see if I understand the logic...
Call OaipmhUtils#listSets to get the newJobSets
Do the four-way comparison above
In the two cases where we don't return from the if clause, then:
get result of getActualNewJobSets (in non-null case) and... getDifference with oldJobSets? newJobSets?

Reply:
If (and only if) oldJobSets is non-empty (i.e. it represents a proper subset of the sets in the repo), then it is possible that sets have been added to the job. In these (two) cases, you must determine the _actual_ new sets (which is a non-empty list):
- if newJobSets is non-empty, then this is simply newJobSets;
- if it _is_ empty (representing a job that should harvest all sets in the repository) then you need to query the OAI-PMH repo to get the list.

Because at least one of the four cases requires performing an async operation (i.e. sending a network request to the repo), the method hasNewSets must also be async.

So if getActualNewSets is of type `Future<List>`, and you need to return `Future<Boolean>` from hasNewSets, then you could do something like this:

```java
return getActualNewSets.map(sets -> {
    // compare with the old sets to determine if "sets" contains elements that are not in "oldJobSets"
    // return a boolean from this lambda
});
```

Does that make sense?